bindings/python/iree/runtime/hal.h

```diff
@@ -115,6 +115,71 @@ class HalBuffer : public ApiRefCounted<HalBuffer, iree_hal_buffer_t> {
   }
 };
 
+// Wrapper around an iree_hal_buffer_mapping_t and iree_hal_buffer_view_t
+// which retains the latter and unmaps/releases on deallocation.
+class HalMappedMemory {
+ public:
+  HalMappedMemory(iree_hal_buffer_mapping_t mapped_memory,
+                  iree_hal_buffer_view_t* bv)
+      : mapped_memory_(mapped_memory), bv_(bv) {
+    iree_hal_buffer_view_retain(bv_);
+  }
+  ~HalMappedMemory() {
+    if (bv_) {
+      iree_hal_buffer_t* buffer = iree_hal_buffer_view_buffer(bv_);
+      iree_hal_buffer_unmap_range(&mapped_memory_);
+      iree_hal_buffer_view_release(bv_);
+    }
+  }
+  HalMappedMemory(HalMappedMemory&& other)
+      : mapped_memory_(other.mapped_memory_), bv_(other.bv_) {
+    other.bv_ = nullptr;
+  }
+
+  static HalMappedMemory Create(HalBufferView& bv) {
+    iree_hal_buffer_t* buffer = iree_hal_buffer_view_buffer(bv.raw_ptr());
+    iree_device_size_t byte_length = iree_hal_buffer_byte_length(buffer);
+    iree_hal_buffer_mapping_t mapped_memory;
+    CheckApiStatus(iree_hal_buffer_map_range(
+                       buffer, IREE_HAL_MEMORY_ACCESS_READ,
+                       0 /* element_offset */, byte_length, &mapped_memory),
+                   "Could not map memory");
+    return HalMappedMemory(mapped_memory, bv.raw_ptr());
+  }
+
+  py::buffer_info ToBufferInfo() {
+    absl::InlinedVector<int32_t, 6> shape(iree_hal_buffer_view_shape_rank(bv_));
+    CheckApiStatus(
+        iree_hal_buffer_view_shape(bv_, shape.size(), shape.data(), nullptr),
+        "Error getting buffer view shape");
+    iree_hal_element_type_t element_type =
+        iree_hal_buffer_view_element_type(bv_);
+    int32_t element_size = iree_hal_element_byte_count(element_type);
+    absl::InlinedVector<py::ssize_t, 6> dims(shape.size());
+    for (int i = 0; i < shape.size(); ++i) {
+      dims[i] = shape[i];
+    }
+    absl::InlinedVector<py::ssize_t, 8> strides(shape.size());
+    if (!strides.empty()) {
+      strides[shape.size() - 1] = element_size;
+      for (int i = shape.size() - 2; i >= 0; --i) {
+        strides[i] = strides[i + 1] * shape[i + 1];
+      }
+    }
+
+    // TODO(laurenzo): We need to figure out how to propagate dtype in the
```

Comment:
don't you do the inverse of this above in `DTYPE_TO_HAL_ELEMENT_TYPE`?

Reply:
Copy-pasta. Just removed the comment: implementing the buffer protocol here is for completeness but not actually used by the bindings (which construct a proper ndarray, which carries these details).